Assets/Scripts/BulletController.cs

```diff
@@ -0,0 +1,70 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using System.Linq;
+public class BulletController : ProjectileController
+{
+    
+    public static void MoveWithGravity(float distance, ref ProjectileState state, GunStats stats)
+    {
+        //Update the position of the projectile
+        state.position += state.direction * distance;
+
+        //Update the velocity of the projectile
+        float time = distance / state.speed;
+        Vector3 velocity = state.direction * state.speed;
+        velocity += Vector3.up * state.gravity * time;
+        state.speed = velocity.magnitude;
+        state.direction = velocity.normalized;
+        
+    }
+
+    public override void Start()
+    {
+        base.Start();
+        UpdateProjectileMovement += MoveWithGravity;
+    }
+
+    private void FixedUpdate()
+    {
+        state.lastUpdateTime = Time.time;
+        state.olderPosition = state.oldPosition;
+        state.oldPosition = state.position;
+
+        Collider[] collisions;
+
+        UpdateProjectileMovement?.Invoke(state.speed * Time.fixedDeltaTime, ref state, stats);
+
+        var direction = state.position - state.oldPosition;
+        RaycastHit[] rayCasts;
+        if (stats.ProjectileSize > 0)
+        {
+            // Both of these need to be ordered
+            rayCasts = Physics.SphereCastAll(transform.position, stats.ProjectileSize, direction, direction.magnitude);
+        }
+        else
+        {
+            // Both of these need to be ordered
+            rayCasts = Physics.RaycastAll(transform.position, direction, direction.magnitude);
+        }
+
+        collisions = rayCasts.OrderBy(x => x.distance).Select(x => x.collider).ToArray();
+        foreach(var collision in collisions)
+        {
+
+            OnColliderHit?.Invoke(collision, ref state, stats);
+            HitboxController hitbox = collision.GetComponent<HitboxController>();
+
+            if (hitbox != null){
+                OnHitboxCollision?.Invoke(hitbox, ref state, stats);
+            }
+        }
```

Comment:
As is, the colliders don't need to be stored in an array. You can simply pass them on as an `IEnumerable<Collider>` out of the `Select`-call:

I would also argue that the hit info could be interesting to pass on to the various events, in case the events could use them for further effects... 🤔 

i.e:
```cs

public delegate void CollisionEvent(RaycastHit hit, ref ProjectileState state, GunStats stats);

//...

foreach(var raycastHit in rayCasts.OrderBy(x => x.distance))
{
    OnColliderHit?.Invoke(raycastHit, ref state, stats);
    HitboxController hitbox = raycastHit.collider.GetComponent<HitboxController>();

    if (hitbox != null)
    {
        OnHitboxCollision?.Invoke(hitbox, ref state, stats);
    }
}
```

Reply:
Not all oncolliderhits will use raycast, some will probably use hitboxes and rigidbodies, if for instance we want grenades